lib/Auth/Process/EnsureVoMember.php

```diff
@@ -0,0 +1,170 @@
+            );
+        }
+
+        $this->loginUrlAttr = $config->getString(self::LOGIN_URL, null);
+        $this->registrarUrl = $config->getString(self::REGISTRAR_URL, null);
+
+        $interface = $config->getString(self::INTERFACE_PROPNAME, self::RPC);
+        $this->adapter = Adapter::getInstance($interface);
+    }
+
+    public function process(&$request)
+    {
+        if (isset($request['SPMetadata']['entityid'])) {
+            $spEntityId = $request['SPMetadata']['entityid'];
+        } else {
+            throw new Exception(
+                'perun:EnsureVoMember: Cannot find entityID of remote SP. ' .
+                'hint: Do you have this filter in IdP context?'
+            );
+        }
+
+        if (isset($request['perun']['user'])) {
+            $user = $request['perun']['user'];
+        } else {
+            throw new Exception(
+                'perun:EnsureVoMember: ' .
+                'missing mandatory field \'perun.user\' in request.' .
+                'Hint: Did you configured PerunIdentity filter before this filter?'
+            );
+        }
+
+        $facility = $this->adapter->getFacilityByEntityId($spEntityId);
+
+        if ($facility === null) {
+            Logger::debug('perun:EnsureVoMember: skip execution - no facility provided');
+            return;
+        }
+
+        $attrValues = $this->adapter->getFacilityAttributesValues(
+            $facility,
+            [$this->voDefsAttr, $this->triggerAttr, $this->loginUrlAttr]
+        );
+
+        $triggerAttrValue = $attrValues[$this->triggerAttr];
+        if ($triggerAttrValue === null || $triggerAttrValue === false) {
+            Logger::debug(
+                'perun:EnsureVoMember: skip execution - attribute ' . self::TRIGGER_ATTR . ' is null or false'
+            );
+            return;
+        }
+
+        if (empty($attrValues[$this->voDefsAttr])) {
+            Logger::debug(
+                'perun:EnsureVoMember: skip execution - attribute ' . self::VO_DEFS_ATTR . ' has null or no value'
+            );
+            return;
+        }
+        $voShortName = $attrValues[$this->voDefsAttr];
+
+        $canAccess = $this->adapter->isUserInVo($user, $voShortName);
+        if ($canAccess) {
+            Logger::debug('perun:EnsureVoMember: user allowed to continue');
+        } else {
+            $this->redirect($request, $attrValues[$this->loginUrlAttr], $voShortName);
+        }
+    }
+
+    private function redirect($request, $loginUrl, $voShortName)
+    {
+        if (!empty($voShortName) &&
+            !empty($this->registrarUrl) &&
+            $this->adapter->hasRegistrationFormByVoShortName($voShortName)
+        ) {
+            $this->redirectDirectly($loginUrl, $voShortName);
+        } else {
+            $this->redirectUnapproved($request);
+        }
+    }
+
+    private function redirectDirectly($loginUrl, $voShortName)
```

Comment:
Maybe rename the method to "redirectToRegistration" to better explaing what is the method doing.